### PR TITLE
Add TracerFactory

### LINKED ISSUE #30: Split the resolver from the "factory"
It's usually a good practice to have a loader (resolver, in this case) and a factory as different entities. This way, the loader can be extended or replaced and the factories would still work. This is specially important for an SPI, such as this one.

One concrete problem caused by the current approach is that one cannot easily provide a custom implementation of the `TracerResolver`, as current resolver (such as Jaeger's) extends the `TracerResolver` and exposes the tracer only via a `protected` method.

My proposal (pseudo code):

```java
public interface TracerFactory {
    Tracer getTracer();
}

public class TracerResolver {
  public static Tracer resolve() {
    ServiceLoader<TracerFactory> factories = ServiceLoader.load(TracerFactory.class);
    // iterators, disambiguation, ...
    return factories.iterator().next().getTracer();
  }
}

// factory example, very similar to current TracerResolver subclasses
public class JaegerTracerFactory implements TracerFactory {
  public Tracer getTracer() {
    return Configuration.fromEnv().getTracer();
  }
}
```

### PR DESCRIPTION
**Pull request by jpkrohling**
Closes #30 

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

File: opentracing-tracerresolver/src/main/java/io/opentracing/contrib/tracerresolver/TracerResolver.java [hunk 1]
```diff
@@ -44,16 +44,33 @@
...
+     *
      * <p>
      * If a {@code GlobalTracer} has been previously registered, it will be returned before attempting to resolve
      * a {@linkplain Tracer} on our own.
+     *
      * <p>
-     * If there are more than one resolver, the first non-<code>null</code> resolved tracer is returned.
+     * If more than one {@link TracerFactory} or {@link TracerResolver} is found, the one with the highest priority is
+     * returned. Note that if a {@link TracerResolver} has a higher priority as all available {@link TracerFactory},
+     * the factory still wins.
```
Comment (sjoerdtalsma) [inline opentracing-tracerresolver/src/main/java/io/opentracing/contrib/tracerresolver/TracerResolver.java]: >  Note that if a {@link TracerResolver} has a higher priority as all available {@link TracerFactory}, the factory still wins.

nit: _higher than_  
:+1: I like this addition, makes the overall priority abundantly clear.

File: opentracing-tracerresolver/src/test/java/io/opentracing/contrib/tracerresolver/TracerResolverTest.java [hunk 2]
```diff
@@ -103,6 +104,19 @@ public void testResolveWithExistingGlobalTracer() throws IOException {
...
     }
 
+    @Test
+    public void testResolveFromFactory() throws IOException {
+        writeServiceFile(TracerFactory.class, Mocks.Prio0_TracerFactory.class);
+        assertThat(TracerResolver.resolveTracer(), is(instanceOf(Mocks.ResolvedTracerFromFactory.class)));
+    }
+
+    @Test
+    public void testResolveFromFactoryHasHigherPriorityAsResolver() throws IOException {
```
Comment (sjoerdtalsma) [inline opentracing-tracerresolver/src/test/java/io/opentracing/contrib/tracerresolver/TracerResolverTest.java]: nit: higher than

**Review by sjoerdtalsma — APPROVED** [review]
Looks good to me and obviously is backwards-compatible

**Comment by jpkrohling** [comment]
PR updated

**Comment by sjoerdtalsma** [comment]
All checks pass; I'll merge and release.
Thanks @jpkrohling  :+1:

**Pull request merged**